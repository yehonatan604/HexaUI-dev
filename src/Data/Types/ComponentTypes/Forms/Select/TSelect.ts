import { TComponent, ComponentOptions } from "../../TComponent";

export type TSelect = TComponent<"select"> & {
    options?: SelectOptions;
    items: TSelectOption[];
}

export type SelectOptions = ComponentOptions & {
    listItemsVariant?: string;
    listItemsBgVariant?: string;
}

export type TSelectOption = {
    value: string;
    label: string;
};
