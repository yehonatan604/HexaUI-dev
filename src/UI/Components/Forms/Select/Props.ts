import { TComponent } from "../../../../Data/Types/ComponentTypes/TComponent";
import { TSelectOption } from "../../../../Data/Types/ComponentTypes/Forms/Select/TSelectOption";

export type SelectProps = TComponent<"select"> & {
    options: TSelectOption[];
    variant?: string;
    listItemsVariant?: string;
};