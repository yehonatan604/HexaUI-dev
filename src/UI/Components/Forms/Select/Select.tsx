import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";
import { TSelect } from "./types/TSelect";

const Select = (props: TSelect) => {
  const { options, items, ...componentProps } = props;
  // Hooks
  const { colors, mode } = useTheme();

  // Options
  const textColor =
    options?.textVariant ||
    (mode === "light" ? colors.standradDark : colors.standradLight);
  const borderColor = options?.border?.variant || colors.standrad;
  const listItemColor =
    options?.listItemsVariant ||
    (mode === "light" ? colors.standradDark : colors.standradLight);
  const listItemsBgColor =
    options?.listItemsBgVariant ||
    (mode === "light" ? colors.backgroundLight : colors.backgroundDark);
  const bgColor = options?.bgVariant || "transparent";

  // JSX
  return (
    <select
      {...componentProps}
      className={`rounded border border-${borderColor} bg-${bgColor} text-${textColor}`}
    >
      {items.map((option, index) => (
        <option
          key={index}
          value={option.value}
          className={`text-${listItemColor} bg-${listItemsBgColor}`}
        >
          {option.label}
        </option>
      ))}
    </select>
  );
};

export default Select;
