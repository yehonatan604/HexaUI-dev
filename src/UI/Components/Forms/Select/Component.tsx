import { getRandomColor } from "../../../../Core/Helpers/Color.helper";
import useTheme from "../../../../Core/Hooks/useTheme";
import { Variants } from "../../../../Data/Constants/Variants";
import { SelectProps } from "./Props";

const Select = (props: SelectProps) => {
  const {
    options,
    variant = Variants.Standard,
    listItemsVariant = Variants.Standard,
    ...componentProps
  } = props;
  const mode = useTheme().mode;

  const textColor =
    variant === Variants.Failure
      ? "text-red-500"
      : variant === Variants.Success
      ? "text-green-500"
      : variant === Variants.Warning
      ? "text-yellow-500"
      : variant === Variants.Info
      ? "text-blue-500"
      : variant === Variants.Lime
      ? "text-lime-500"
      : variant === Variants.Cyan
      ? "text-cyan-500"
      : variant === Variants.Purple
      ? "text-purple-500"
      : variant === Variants.Pink
      ? "text-pink-500"
      : mode === "light"
      ? "text-zinc-800"
      : "text-zinc-300";
  const bgColor = mode === "light" ? "bg-zinc-100" : "bg-zinc-800";

  const borderColor =
    variant === Variants.Failure
      ? "border-red-500"
      : variant === Variants.Success
      ? "border-green-500"
      : variant === Variants.Warning
      ? "border-yellow-500"
      : variant === Variants.Info
      ? "border-blue-500"
      : variant === Variants.Lime
      ? "border-lime-500"
      : variant === Variants.Cyan
      ? "border-cyan-500"
      : variant === Variants.Purple
      ? "border-purple-500"
      : variant === Variants.Pink
      ? "border-pink-500"
      : mode === "light"
      ? "border-zinc-800"
      : "border-zinc-300";

  const listItemColor =
    listItemsVariant === Variants.Failure
      ? "text-red-500"
      : listItemsVariant === Variants.Success
      ? "text-green-500"
      : listItemsVariant === Variants.Warning
      ? "text-yellow-500"
      : listItemsVariant === Variants.Info
      ? "text-blue-500"
      : listItemsVariant === Variants.Lime
      ? "text-lime-500"
      : listItemsVariant === Variants.Cyan
      ? "text-cyan-500"
      : listItemsVariant === Variants.Purple
      ? "text-purple-500"
      : listItemsVariant === Variants.Pink
      ? "text-pink-500"
      : listItemsVariant === "random"
      ? getRandomColor()
      : mode === "light"
      ? "text-zinc-800"
      : "text-zinc-300";

  return (
    <select
      {...componentProps}
      className={`rounded border ${borderColor} ${textColor} ${bgColor} cursor-pointer`}
    >
      {options.map((option, index) => (
        <option
          key={index}
          value={option.value}
          className={`${
            listItemsVariant
              ? listItemsVariant === "rainbow"
                ? getRandomColor()
                : listItemColor
              : textColor
          }`}
        >
          {option.label}
        </option>
      ))}
    </select>
  );
};

export default Select;
