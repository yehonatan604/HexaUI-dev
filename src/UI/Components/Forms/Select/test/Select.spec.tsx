import { render, screen, fireEvent } from "@testing-library/react";
import Select from "../Select"; // Adjust the import path
import { describe, it, expect } from "vitest";

describe("Select Component", () => {
  const items = [
    { value: "1", label: "Option 1" },
    { value: "2", label: "Option 2" },
    { value: "3", label: "Option 3" },
  ];

  it("should render the select element with the provided items", () => {
    render(<Select items={items} />);

    // Check if the select element is rendered
    const selectElement = screen.getByRole("combobox");
    expect(selectElement).toBeInTheDocument();

    // Check if all the options are rendered
    items.forEach((item) => {
      expect(screen.getByText(item.label)).toBeInTheDocument();
    });
  });

  it("should update the selected option when changed", () => {
    render(<Select items={items} />);

    // Cast to HTMLSelectElement to access the value property
    const selectElement = screen.getByRole("combobox") as HTMLSelectElement;

    // Initially, no option should be selected
    expect(selectElement.value).toBe("1");

    // Select a new option
    fireEvent.change(selectElement, { target: { value: "2" } });

    // The selected value should update
    expect(selectElement.value).toBe("2");
  });

  it("should apply custom text and background colors based on options", () => {
    render(
      <Select
        items={items}
        options={{
          textVariant: "green-500",
          bgVariant: "blue-100",
          border: { variant: "red-500" },
          listItemsVariant: "yellow-500",
          listItemsBgVariant: "gray-300",
        }}
      />
    );

    const selectElement = screen.getByRole("combobox");

    // Check if the correct text, background, and border classes are applied to the select element
    expect(selectElement).toHaveClass("text-green-500", "bg-blue-100", "border-red-500");

    // Check if the correct text and background colors are applied to the list items
    const optionElement = screen.getByText("Option 1");
    expect(optionElement).toHaveClass("text-yellow-500", "bg-gray-300");
  });

  it("should render and match the snapshot", () => {
    const { asFragment } = render(<Select items={items} />);

    // Snapshot the select component structure
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with custom styles and match the snapshot", () => {
    const { asFragment } = render(
      <Select
        items={items}
        options={{
          textVariant: "green-500",
          bgVariant: "blue-100",
          border: { variant: "red-500" },
          listItemsVariant: "yellow-500",
          listItemsBgVariant: "gray-300",
        }}
      />
    );

    // Snapshot the select component structure with custom styles
    expect(asFragment()).toMatchSnapshot();
  });
});
