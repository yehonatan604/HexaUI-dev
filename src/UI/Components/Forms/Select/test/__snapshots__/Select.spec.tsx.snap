// Vitest Snapshot v1

exports[`Select Component > should render and match the snapshot 1`] = `
<DocumentFragment>
  <select
    class="rounded border border-border-gray-300 dark:border-gray-600 bg-bg-transparent text-text-gray-800 dark:text-gray-200"
  >
    <option
      class="text-text-gray-800 dark:text-gray-200 bg-bg-white dark:bg-gray-800"
      value="1"
    >
      Option 1
    </option>
    <option
      class="text-text-gray-800 dark:text-gray-200 bg-bg-white dark:bg-gray-800"
      value="2"
    >
      Option 2
    </option>
    <option
      class="text-text-gray-800 dark:text-gray-200 bg-bg-white dark:bg-gray-800"
      value="3"
    >
      Option 3
    </option>
  </select>
</DocumentFragment>
`;

exports[`Select Component > should render with custom styles and match the snapshot 1`] = `
<DocumentFragment>
  <select
    class="rounded border border-red-500 bg-blue-100 text-green-500"
  >
    <option
      class="text-yellow-500 bg-gray-300"
      value="1"
    >
      Option 1
    </option>
    <option
      class="text-yellow-500 bg-gray-300"
      value="2"
    >
      Option 2
    </option>
    <option
      class="text-yellow-500 bg-gray-300"
      value="3"
    >
      Option 3
    </option>
  </select>
</DocumentFragment>
`;
